File: week13/11/1/P1.cpp
#include <iostream>
#include <vector>
#include <queue>

using namespace std;

class Solution {
public:
    int lastStoneWeight(vector<int>& stones) {
        // 이렇게 해야 min-heap이 됨
        priority_queue<int> stone;
        for(auto st: stones){
            stone.push(st);
        }
        int big;
        int mid;
        while(stone.size() >= 2){
            big = stone.top();
            stone.pop();
            mid = stone.top();
            stone.pop();
            if(big > mid){
                stone.push(big - mid);
            }
        }
        if (stone.size() == 1){
            return stone.top();
        }
        else{
            return 0;
        }
    }
};

File: week13/11/1/P1.py
class Solution:
    def lastStoneWeight(self, stones: List[int]) -> int:
        stones.sort()
        while len(stones) >= 2:
            big = stones.pop()
            mid = stones.pop()
            if big > mid:
                stones.append(big - mid)
                stones.sort()
            else:
                stones.sort()
        
        if len(stones) == 1:
            return stones[0]
        else:
            return 0
